### PR TITLE
Fix SG parser running on tesseract 5.x

### PR DESCRIPTION
**Pull request by r24mille**
# tl;dr

This PR fixes the tesseract datetime detection bugs I have been experiencing locally with the SG parser. It also updates the SG zone's solar output capacity.

# The Problem
When running the unit tests locally, I get the following output:

```bash
$ poetry run test
⚙︎ python -u -m unittest discover tests 
./REDACTED/electricitymap-contrib-venv/lib/python3.8/site-packages/html5lib/_trie/_base.py:3: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
  from collections import Mapping
....
----------------------------------------------------------------------
Ran 5 tests in 0.308s

OK
⚙︎ python -u -m unittest discover parsers/test 
/REDACTED/electricitymap-contrib-venv/lib/python3.8/site-packages/bs4/element.py:1615: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
  if (isinstance(value, str) or isinstance(value, collections.Callable) or hasattr(value, 'match')
...................................................F..F..........................
======================================================================
FAIL: test_allow_remote_clock_to_be_slightly_ahead (test_SG.TestSolar)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/REDACTED/electricitymap-contrib-venv/lib/python3.8/site-packages/freezegun/api.py", line 735, in wrapper
    result = func(*args, **kwargs)
  File "/REDACTED/electricitymap-contrib/parsers/test/test_SG.py", line 44, in test_allow_remote_clock_to_be_slightly_ahead
    self.assertEqual(power, 0.0)
AssertionError: None != 0.0

======================================================================
FAIL: test_works_when_zero (test_SG.TestSolar)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/REDACTED/electricitymap-contrib-venv/lib/python3.8/site-packages/freezegun/api.py", line 735, in wrapper
    result = func(*args, **kwargs)
  File "/REDACTED/electricitymap-contrib/parsers/test/test_SG.py", line 32, in test_works_when_zero
    self.assertEqual(power, 0.0)
AssertionError: None != 0.0

----------------------------------------------------------------------
Ran 81 tests in 49.059s

FAILED (failures=2)
FAILED: python -u -m unittest discover parsers/test
```

# Background / Investigation
I have the following `tesseract` version installed locally.

```bash
$ tesseract --version
tesseract 5.0.0
 leptonica-1.79.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 2.0.3) : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.1
 Found AVX2
 Found AVX
 Found FMA
 Found SSE4.1
 Found OpenMP 201511
 Found libarchive 3.4.0 zlib/1.2.11 liblzma/5.2.4 bz2lib/1.0.8 liblz4/1.9.2 libzstd/1.4.4
```

When investigating with a debugger, the culprit is datetime detection in the SG solar output image when processed by tesseract. The third line, which is the datetime line, result in gibberish character recognition.

datetime-gibberish

I think this may be caused by me running a recent version of tesseract locally. While investigating these code changes, [light text on dark backgrounds seemed to be a problem](https://stackoverflow.com/q/54782568) with the recent changes to tesseract's OCR engine modes.

# The Fixes
This is fixed predominantly with the dark text on light background pre-processing in eb85e7104056c035b29bd3f5e1b9837eb06f2230 and 7b15102db261f260517ffe3aec304a5b0fbe9c57.

I also noticed that the Singapore solar capacity changed from 231.26 MW to 461.66 MW. See https://www.ema.gov.sg/cmsmedia/irradiance/plot.png for details.

plot

### CODE REVIEW

File: parsers/SG.py [hunk 1]
```diff
@@ -267,6 +232,79 @@ def fetch_price(zone_key='SG', session=None, target_datetime=None, logger=None)
...
+
+def __detect_output_from_solar_image(solar_image, logger):
+    w, h = solar_image.size
+    crop_left = int(w * 0.65)
+    crop_top = int(h * 0.74)
+    crop_right = int(w * 0.93)
+    crop_bottom = int(h * 0.80)
+    output_img = solar_image.crop((crop_left, crop_top, crop_right, crop_bottom))
+    processed_img = __preprocess_image_for_ocr(output_img)
+    text = image_to_string(processed_img, lang='eng', config='--psm 7')
```
Comment (r24mille) [inline parsers/SG.py:265]: Originally, I added the [page segmentation method](https://tesseract-ocr.github.io/tessdoc/ImproveQuality#page-segmentation-method) config to try and improve the "OMWac" issue in the middle of the night when the decimal place vanishes from the PV output.

It didn't really help much, but I've left it in, since I think `--psm 7` is most descriptive of the changes I made to cropping the output line and datetime lines separately, "Treat the image as a single text line."

I also played around with `tessedit_char_whitelist` but that did not improve the "OMWac" issue. I really wish we could use `user-patterns` to restrict character recognition to a regex-like pattern. However, [user-patterns can only be passed in via a file reference](https://tesseract-ocr.github.io/tessdoc/APIExample-user_patterns.html#cli-example). Adding that file and accounting for all the different relative paths that may appear felt like too much complexity for a scenario that's easily handled by the existing `try`/`except` block below.

File: parsers/SG.py [hunk 2]
```diff
@@ -267,6 +232,79 @@ def fetch_price(zone_key='SG', session=None, target_datetime=None, logger=None)
...
 
+def __detect_datetime_from_solar_image(solar_image, logger):
+    w, h = solar_image.size
+    crop_left = int(w * 0.75)
+    crop_top = int(h * 0.87)
+    crop_right = int(w * 0.93)
+    crop_bottom = int(h * 0.92)
+    time_img = solar_image.crop((crop_left, crop_top, crop_right, crop_bottom))
+    processed_img = __preprocess_image_for_ocr(time_img)
+    text = image_to_string(processed_img, lang='eng', config='--psm 7 -c tessedit_char_whitelist="0123456789:- "')
```
Comment (r24mille) [inline parsers/SG.py:243]: Similar to [my other comment](https://github.com/electricityMap/electricitymap-contrib/pull/3620/files#r774752757), [page segmentation method](https://tesseract-ocr.github.io/tessdoc/ImproveQuality#page-segmentation-method) 7 is most descriptive of the more specific image cropping that I split things into.

Here, `tessedit_char_whitelist` did improve the accuracy of datetime values detected in the image.

File: parsers/SG.py [hunk 3]
```diff
@@ -267,6 +232,79 @@ def fetch_price(zone_key='SG', session=None, target_datetime=None, logger=None)
...
 
 
+def __detect_datetime_from_solar_image(solar_image, logger):
+    w, h = solar_image.size
+    crop_left = int(w * 0.75)
+    crop_top = int(h * 0.87)
+    crop_right = int(w * 0.93)
+    crop_bottom = int(h * 0.92)
+    time_img = solar_image.crop((crop_left, crop_top, crop_right, crop_bottom))
+    processed_img = __preprocess_image_for_ocr(time_img)
```
Comment (r24mille) [inline parsers/SG.py:242]: Result after new cropping and pre-processing technique is:
pv-datetime-preprocess-result

File: parsers/SG.py [hunk 4]
```diff
@@ -267,6 +232,79 @@ def fetch_price(zone_key='SG', session=None, target_datetime=None, logger=None)
...
+
+
+def __detect_output_from_solar_image(solar_image, logger):
+    w, h = solar_image.size
+    crop_left = int(w * 0.65)
+    crop_top = int(h * 0.74)
+    crop_right = int(w * 0.93)
+    crop_bottom = int(h * 0.80)
+    output_img = solar_image.crop((crop_left, crop_top, crop_right, crop_bottom))
+    processed_img = __preprocess_image_for_ocr(output_img)
```
Comment (r24mille) [inline parsers/SG.py:264]: Result after new cropping and pre-processing technique is:
est-output-preprocess-result

**Review by madsnedergaard — APPROVED** [review]
Thanks for taking the time to fix this! 💪 
It still works on my machine (`tesseract v4.1.1`)  and as far as I can tell, Drone is [also using v4 (through `tesseract-ocr` package)](https://app.circleci.com/pipelines/github/electricityMap/electricitymap-contrib/3227/workflows/64dc9a24-d6e1-4bfd-87af-93240bb2018e/jobs/19364/parallel-runs/0/steps/0-104).

And in production it seems like we're using an oooold version of `pytesseract`, but let's try merging and see if it still works :)  https://github.com/electricityMap/electricitymap-contrib/blob/e9c26e3b333e877ba42b115f74850de8a7ec75d0/poetry.lock#L377-L383
